src/UpgradeRule/AbstractUpgradeRule.php

```diff
@@ -15,6 +17,26 @@
     protected $warningCollector = [];
 
     /**
+     * Called on a code collection prior to upgrade
+     *
+     * @param CollectionInterface $code
+     * @param CodeChangeSet $changeset
+     */
+    public function beforeUpgrade(CollectionInterface $code, CodeChangeSet $changeset)
```

Comment:
Maybe `beforeUpgradeCollection()` and `afterUpgradeCollection()` are better names? If someone was upgrading a single file they might not call this.